File: 2015/python/day17.py
import fileinput


def partitions(total, buckets):
    def partitions_recurse(total, min_ix):
        if total == 0:
            yield []
        elif total < 0:
            raise StopIteration
        else:
            for i, bucket in enumerate(buckets[min_ix:]):
                for partition in partitions_recurse(total - bucket, min_ix + i + 1):
                        yield [bucket] + partition

    return partitions_recurse(total, 0)


def part1(buckets):
    parts = list(partitions(150, buckets))

    return len(parts)


def part2(buckets):
    parts = list(partitions(150, buckets))
    min_buckets = min(len(p) for p in parts)

    return sum(1 for p in parts if len(p) == min_buckets)


if __name__ == "__main__":
    lines = [line.strip() for line in fileinput.input()]
    buckets = [int(line) for line in lines]

    print("Part 1: %s" % part1(buckets))
    print("Part 2: %s" % part2(buckets))

File: 2015/rust/src/bin/day17.rs
use anyhow::{anyhow, Result};

use aoc::utils::{get_input, parse_fields};

type Volume = u64;

fn main() -> anyhow::Result<()> {
    let input = get_input()?;
    let buckets: Vec<Volume> = parse_fields(input.trim(), '\n')?;

    println!("Part 1: {}", part1(&buckets));
    println!("Part 2: {}", part2(&buckets)?);

    Ok(())
}

fn part1(buckets: &[Volume]) -> usize {
    partitions(150, buckets).len()
}

fn part2(buckets: &[Volume]) -> Result<usize> {
    let parts = partitions(150, buckets);
    let lengths: Vec<usize> = parts.iter().map(|p| p.len()).collect();
    let min_length = lengths.iter().min().ok_or(anyhow!("No buckets."))?;

    Ok(lengths.iter().filter(|&l| l == min_length).count())
}

fn partitions(total: Volume, buckets: &[Volume]) -> Vec<Vec<Volume>> {
    if total == 0 {
        return vec![vec![]];
    }
    let mut results = vec![];
    for (i, &bucket) in buckets.iter().enumerate() {
        if total >= bucket {
            for mut partition in partitions(total - bucket, &buckets[i + 1..]) {
                partition.push(bucket);
                results.push(partition);
            }
        }
    }
    results
}

#[cfg(test)]
mod tests {
    use super::partitions;
    use aoc::testing::assert_matches_ignoring_order;

    #[test]
    fn all_partitions() {
        let result = partitions(25, &[20, 15, 10, 5, 5]);
        let expected = vec![vec![5, 20], vec![5, 20], vec![10, 15], vec![5, 5, 15]];

        assert_matches_ignoring_order(&result, &expected)
    }
}
